R/dummy_extract.R

```diff
@@ -0,0 +1,280 @@
+#'     names()
+#'
+#' tidy(dummies, number = 1)
+#' tidy(dummies_specific, number = 1)
+#'
+#' # pattern argument can be useful to extract harder patterns
+#' color_examples <- tibble(
+#'   colors = c("['red', 'blue']",
+#'              "['red', 'blue', 'white']",
+#'              "['blue', 'blue', 'blue']")
+#' )
+#'
+#' dummies_color <- recipe(~ colors, data = color_examples) %>%
+#'   step_dummy_extract(colors, pattern = "(?<=')[^',]+(?=')") %>%
+#'   prep()
+#'
+#' dommies_data_color <- dummies_color %>%
+#'   bake(new_data = NULL)
+#'
+#' dommies_data_color
+step_dummy_extract <-
+  function(recipe,
+           ...,
+           role = "predictor",
+           trained = FALSE,
+           sep = NULL,
+           pattern = NULL,
+           threshold = 0.0,
+           other = "other",
+           naming = dummy_names,
+           levels = NULL,
+           keep_original_cols = FALSE,
+           skip = FALSE,
+           id = rand_id("dummy_extract")) {
+
+    if (!is_tune(threshold) & !is_varying(threshold)) {
+      if (threshold < 0) {
+        rlang::abort("`threshold` should not be negative.")
+      }
+      if (threshold >= 1 && !is_integerish(threshold)) {
+        rlang::abort("If `threshold` is greater than one it should be an integer.")
+      }
+    }
+
+    add_step(
+      recipe,
+      step_dummy_extract_new(
+        terms = enquos(...),
+        role = role,
+        trained = trained,
+        sep = sep,
+        pattern = pattern,
+        threshold = threshold,
+        other = other,
+        naming = naming,
+        levels = levels,
+        keep_original_cols = keep_original_cols,
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+step_dummy_extract_new <-
+  function(terms, role, trained, sep, pattern, threshold, other, naming, levels,
+           keep_original_cols, skip, id) {
+    step(
+      subclass = "dummy_extract",
+      terms = terms,
+      role = role,
+      trained = trained,
+      sep = sep,
+      pattern = pattern,
+      threshold = threshold,
+      other = other,
+      naming = naming,
+      levels = levels,
+      keep_original_cols = keep_original_cols,
+      skip = skip,
+      id = id
+    )
+  }
+
+#' @export
+prep.step_dummy_extract <- function(x, training, info = NULL, ...) {
+  col_names <- recipes_eval_select(x$terms, training, info)
+
+  if (length(col_names) > 0) {
+    col_names <- check_factor_vars(training, col_names, "step_dummy")
+
+    ## I hate doing this but currently we are going to have
+    ## to save the terms object from the original (= training)
+    ## data
+    levels <- vector(mode = "list", length = length(col_names))
+    names(levels) <- col_names
+    for (col_name in col_names) {
+      elements <- dummy_extract_extract(
+        training[[col_name]], sep = x$sep, pattern = x$pattern
+      )
+
+      lvls <- map(elements, unique)
+      lvls <- unlist(lvls)
+      lvls <- sort(table(lvls), decreasing = TRUE)
```

Comment:
Please make this work for proportions and counts as mentioned ☝️

Reply:
Done!